Differential Binarization model

mehtamansi29 opened the pull request:
## Description of the Change
This PR adds Real-time Scene Text Detection with Differentiable Binarization model, Differential Binarization Model(diffbin) to KerasHub. 


## Reference
For details, see Real-time Scene Text Detection with Differentiable Binarization model ([Paper](https://arxiv.org/pdf/1911.08947))

## Colab Notebook
Usage and weight transfer [Colab](https://colab.sandbox.google.com/gist/mehtamansi29/ea4ad53f43626e2415a332034ad9e0f7/diffbin_model_final.ipynb)

## Checklist
- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub Model contribution guidelines in making these changes.
- [x] I have followed the Keras Hub API design guidelines in making these changes.
- [x] I have signed the Contributor License Agreement.

---

mehtamansi29 committed:
ImageText detector preprocessor for Differential Binarization model

sineeli commented on keras_hub/src/models/image_text_detector_preprocessor.py:
```diff
@@ -0,0 +1,62 @@
+            output = self.image_converter(
+                {
+                    "images": x,
+                    "bounding_boxes": y,
```
I don't think we need any bounding boxes for this task.

mehtamansi29 committed:
db_utils functions and testfile

mehtamansi29 committed:
Diffbin utils function and test file

mehtamansi29 committed:
diffbin utils function and testfile

mehtamansi29 committed:
diffbin preprocessing function

mehtamansi29 committed:
diffbin postprocessing function

mehtamansi29 committed:
diffbin postprocessing function_1

mehtamansi29 committed:
diffbin postprocessing function_2

mehtamansi29 committed:
diffbin postprocessing function_3

sachinprasadhs reviewed (commented):
Took high level pass and left some comments.
Also,
Make al the file names in follow the same format like other files, for db_utils and losses.py

sachinprasadhs commented on keras_hub/src/models/image_text_detector_preprocessor.py:
```diff
@@ -0,0 +1,71 @@
+            mask= get_mask(width,height,poly)
+            return keras.utils.pack_x_y_sample_weight(x,mask, sample_weight)
+    
+    @preprocessing_function
+    def generate_postprocess(self,x):
+        '''
+        Generates postprocess function to convert probability map of 
+        model output to polygon
+        '''
+        probability_maps,threshold_maps = x["probability_maps"], x["threshold_maps"]
+        binary_maps = 1.0 / (1.0 + keras.ops.exp(-50.0 * (probability_maps - threshold_maps)))
+        outputs = keras.layers.Concatenate(axis=-1)(
+            [probability_maps, threshold_maps, binary_maps])
+        return outputs
+        
```
why do we need this? we don't have `.generate` function for diff_bin right?

mehtamansi29 replied:
Here model output is binary_maps and we need to obtain to obtain a polygon representation. 
Earlier it was in difbin_textdetector.py file. 
But now as we are creating preprocess and postprocess  in image_text_detector_preprocessor.py, so as per the design of kerashub models.

sachinprasadhs commented on keras_hub/src/models/image_text_detector_preprocessor.py:
```diff
@@ -0,0 +1,71 @@
+    def __init__(
+        self,
+        image_converter=None,
+        target_size=(640, 640),
+        shrink_ratio=0.3,
```
This will be image_size and will be handled in ImageConverter, not in this

mehtamansi29 replied:
I need to use image_size for getting width and height generating mask input image.

mehtamansi29 committed:
Merge branch 'keras-team:master' into diffbin

mehtamansi29 committed:
diffbin preocessing and db_utils completed

mehtamansi29 committed:
Merge branch 'keras-team:master' into diffbin

mehtamansi29 committed:
diffbin_backbone model creation and backboone test for diffbin segmented output

mehtamansi29 committed:
Merge branch 'keras-team:master' into diffbin

mehtamansi29 committed:
modifited diffbin _textdetector

mehtamansi29 committed:
Updates image_text_detector preprocessor

mehtamansi29 committed:
Updates image_text_detector preprocessor with ignores argument

mehtamansi29 committed:
Updates image_text_detector preprocessor,db_utils and formatting with shell_format

mehtamansi29 committed:
Updates image_text_detector_1

mehtamansi29 committed:
Updates image_text_detector_1

mehtamansi29 committed:
Updates image_text_detector_3

mehtamansi29 committed:
Updates image_text_detector_3

mehtamansi29 committed:
Updates image_text_detector_4

mehtamansi29 committed:
Updates image_text_detector_5

mehtamansi29 committed:
Updates image_text_detector_6

mehtamansi29 committed:
Updates image_text_detector_7

mehtamansi29 committed:
annotation size

mehtamansi29 committed:
Merge branch 'keras-team:master' into diffbin

mehtamansi29 committed:
fill poly keras chages

mehtamansi29 committed:
fill poly keras changes revert

mehtamansi29 committed:
diffbin_imagetextdetector import changes

mehtamansi29 committed:
diffbin_imagetextdetector  changes

mehtamansi29 committed:
diffbin_imagetextdetector  and precommit changes

mehtamansi29 committed:
diffbin_imagetextdetector and precommit changes

mehtamansi29 committed:
diffbin_imagetextdetector and precommit changes

mehtamansi29 committed:
diffbin_textdetector_1

mehtamansi29 committed:
diffbin_textdetector_2

mehtamansi29 committed:
diffbin_textdetector_3

mehtamansi29 committed:
diffbin_textdetector_4

mehtamansi29 committed:
Merge branch 'keras-team:master' into diffbin

mehtamansi29 committed:
diffbin_backbon_image_shape

hertschuh committed:
Fix PaliGemmaCausalLM example. (#2302)

The inputs to `generate` are `"prompts"`, not `"text"`.

Fixes https://github.com/keras-team/keras-hub/issues/1685

divyashreepathihalli committed:
Routine HF sync (#2303)

* routine HF sync

* code reformat

laxmareddyp committed:
incorrect condition on self.sliding_window_size (#2289)

dependabot[bot] committed:
Bump the python group with 2 updates (#2282)

Bumps the python group with 2 updates: torch and torchvision.


Updates `torch` from 2.6.0+cu126 to 2.7.0+cu126

Updates `torchvision` from 0.21.0+cu126 to 0.22.0+cu126

---
updated-dependencies:
- dependency-name: torch
  dependency-version: 2.7.0+cu126
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: python
- dependency-name: torchvision
  dependency-version: 0.22.0+cu126
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: python
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

sonali-kumari1 committed:
Modify TransformerEncoder masking documentation (#2297)

* Modify TransformerEncoder masking documentation

* Added space before parenthesis

pctablet505 committed:
Fix Gemma3InterleaveEmbeddings JAX inference error by ensuring indices are int32 (#2305)

laxmareddyp committed:
update preset versions (#2307)

laxmareddyp committed:
Fix Mistral conversion script (#2306)

* Fix Mistral conversion script

This commit addresses several issues in the Mistral checkpoint conversion script:

- Adds `dropout` to the model initialization to match the Hugging Face model.
- Replaces `requests.get` with `hf_hub_download` for more reliable tokenizer downloads.
- Adds support for both `tokenizer.model` and `tokenizer.json` to handle different Mistral versions.
- Fixes a `TypeError` in the `save_to_preset` function call.

* address format issues

* adopted to latest hub style

* address format issues

---------

Co-authored-by: laxmareddyp <laxmareddyp@laxma-n2-highmem-256gbram.us-central1-f.c.gtech-rmi-dev.internal>

mehtamansi29 committed:
diffbin_backbon_image_preprocessor

mehtamansi29 committed:
diffbin textdetector update

mehtamansi29 committed:
diffbin textdetector update_1

mehtamansi29 committed:
new loss function added

mehtamansi29 committed:
loss updated in diffbin_textdetector

mehtamansi29 committed:
loss updated in diffbin_textdetector_1

mehtamansi29 committed:
loss updated in diffbin_textdetector_2

mehtamansi29 committed:
diffbi_text_detector_update_1

mehtamansi29 committed:
Update DB loss function

mehtamansi29 committed:
Update DB loss function_1

mehtamansi29 committed:
Update DB loss function_2

mehtamansi29 committed:
Update DB loss function_3

mehtamansi29 committed:
Update DB loss function_4

mehtamansi29 committed:
Update DB loss function_5

mehtamansi29 committed:
Update DB loss function_6

mehtamansi29 committed:
Update DB loss function_7

mehtamansi29 committed:
Update DB loss function_8

mehtamansi29 committed:
Update DB loss function_9

mehtamansi29 committed:
Update DB loss function_8

dependabot[bot] committed:
Bump the python group with 6 updates (#2317)

Updates the requirements on [tensorflow-cpu](https://github.com/tensorflow/tensorflow), [tensorflow](https://github.com/tensorflow/tensorflow), [tensorflow-text](https://github.com/tensorflow/text), torch, torchvision and [tensorflow[and-cuda]](https://github.com/tensorflow/tensorflow) to permit the latest version.

Updates `tensorflow-cpu` to 2.19.0
- [Release notes](https://github.com/tensorflow/tensorflow/releases)
- [Changelog](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md)
- [Commits](https://github.com/tensorflow/tensorflow/compare/v2.18.1...v2.19.0)

Updates `tensorflow` to 2.19.0
- [Release notes](https://github.com/tensorflow/tensorflow/releases)
- [Changelog](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md)
- [Commits](https://github.com/tensorflow/tensorflow/compare/v2.18.1...v2.19.0)

Updates `tensorflow-text` to 2.19.0
- [Release notes](https://github.com/tensorflow/text/releases)
- [Commits](https://github.com/tensorflow/text/compare/v2.18.0...v2.19.0)

Updates `torch` from 2.7.0+cu126 to 2.7.1+cu126

Updates `torchvision` from 0.22.0+cu126 to 0.22.1+cu126

Updates `tensorflow[and-cuda]` to 2.19.0
- [Release notes](https://github.com/tensorflow/tensorflow/releases)
- [Changelog](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md)
- [Commits](https://github.com/tensorflow/tensorflow/compare/v2.18.0...v2.19.0)

---
updated-dependencies:
- dependency-name: tensorflow-cpu
  dependency-version: 2.19.0
  dependency-type: direct:production
  dependency-group: python
- dependency-name: tensorflow
  dependency-version: 2.19.0
  dependency-type: direct:production
  dependency-group: python
- dependency-name: tensorflow-text
  dependency-version: 2.19.0
  dependency-type: direct:production
  dependency-group: python
- dependency-name: torch
  dependency-version: 2.7.1+cu126
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: python
- dependency-name: torchvision
  dependency-version: 0.22.1+cu126
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: python
- dependency-name: tensorflow[and-cuda]
  dependency-version: 2.19.0
  dependency-type: direct:production
  dependency-group: python
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

kanpuriyanawab committed:
Qwen3 causal lm (#2311)

* init

* update

* bug fixes

* add qwen causal lm test

* fix qwen3 tests

sachinprasadhs committed:
Update JAX GPU version (#2319)

pass-lin committed:
support flash-attn at torch backend (#2257)

* support flash-attn at torch backend

* fix

* fix

* fix

* fix conflit

* fix conflit

* fix conflit

* fix conflit

* fix conflit

* fix conflit

* format

harshaljanjani committed:
Add HGNetV2 to KerasHub (#2293)

* init: Add initial project structure and files

* bug: Small bug related to weight loading in the conversion script

* finalizing: Add TIMM preprocessing layer

* incorporate reviews: Consolidate stage configurations and improve API consistency

* bug: Unexpected argument error in JAX with Keras 3.5

* small addition for the D-FINE to come: No changes to the existing HGNetV2

* D-FINE JIT compile: Remove non-essential conditional statement

* refactor: Address reviews and fix some nits

laxmareddyp committed:
Qwen3 presets register (#2325)

* Register qwen3 presets

* fix format

mehtamansi29 committed:
Update diffbin loss function and test file for loss function_1

mehtamansi29 committed:
Update diffbin loss function and test file for loss function_1

mehtamansi29 committed:
Update diffbin loss function_1

mehtamansi29 committed:
Update diffbin loss function_2

mehtamansi29 committed:
Update diffbin loss function_3

mehtamansi29 committed:
Update diffbin loss function_4

sachinprasadhs commented on keras_hub/src/models/diffbin/diffbin_backbone.py:
```diff
@@ -0,0 +1,220 @@
+    """
+
+    def __init__(
+        self,
+        image_encoder,
+        fpn_channels=256,
+        head_kernel_list=[3, 2, 2],
+        image_shape=(640, 640, 3),
+        dtype=None,
```
you can keep` image_shape=(None, None, 3)` for consistency

mehtamansi29 committed:
Gemini Suggested changes

mehtamansi29 committed:
Resolved conflict

mehtamansi29 committed:
Resolved conflict

mehtamansi29 committed:
Revert "Gemini Suggested changes"

This reverts commit 80b3c56740b8523e8158970d6eb37f41f55142cc.

mehtamansi29 committed:
Resolving Conflicts_1

mehtamansi29 committed:
resolving conflict___1

mehtamansi29 committed:
resolving conflict___2

mehtamansi29 committed:
resolving conflict___3

mehtamansi29 committed:
Merge remote-tracking branch 'upstream/master' into diffbin

mehtamansi29 committed:
Merge branch 'keras-team:master' into diffbin

mehtamansi29 committed:
Resolve backend failing testcases

sachinprasadhs commented on keras_hub/src/models/image_text_detector_preprocessor.py:
```diff
@@ -0,0 +1,57 @@
+
+
+@keras_hub_export("keras_hub.models.ImageTextDetectorPreprocessor")
+class ImageTextDetectorPreprocessor(Preprocessor):
```
I don't see any difference between [ImageClassifierPreprocessor](https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/models/image_classifier_preprocessor.py) and this, if it is used only for preprocessing images, may be you can just use ImageClassifierPreprocessor

mehtamansi29 replied:
This is because earlier I was using diffbin_utils functions(some functions like cv2,shapely created with keras.ops). But this was giving not accurate result as with cv2 and shapely result. So I tried with current approach. 

So @sachinprasadhs Should I remove image_text_detector_preprocessor.py](https://github.com/keras-team/keras-hub/pull/2095/files/56e4e50f176e9f97637e6884b9c530dc5d7b498a#diff-7c22f710ac28ca87452f156f06aaa17c778c613d93764c53870c69768cdb1967) for now and use this [ImageClassifierPreprocessor](https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/models/image_classifier_preprocessor.py)  ?

sachinprasadhs replied:
Then we can subclass `ImageClassifierPreprocessor`, like other vision models

sachinprasadhs reviewed (commented):
Added my review comment, in the issue description add the original implementation, reference paper, training colab and end to end working demo of the trained model including post processing.

mehtamansi29 committed:
Merge branch 'keras-team:master' into diffbin

mehtamansi29 commented on keras_hub/src/models/diffbin/diffbin_textdetector.py:
```diff
@@ -0,0 +1,114 @@
+    and a `keras_hub.models.Preprocessor` to create a model that can be used
+    for detecting text in natural images.
+
+    The probability map output generated by `predict()` can be translated into
+    polygon representation using `model.postprocess_to_polygons()`.
```
Apply the changes with relevant doc string. Next commit able to see the changes

sachinprasadhs commented on keras_hub/src/models/diffbin/diffbin_backbone.py:
```diff
@@ -0,0 +1,218 @@
+        config = super().get_config()
+        config["fpn_channels"] = self.fpn_channels
+        config["head_kernel_list"] = self.head_kernel_list
+        config["image_encoder"] = keras.layers.serialize(self.image_encoder)
```
try `keras.saving.serialize_keras_object`

mehtamansi29 committed:
Merge branch 'keras-team:master' into diffbin

mehtamansi29 committed:
Resolve  failing testcases

mehtamansi29 committed:
Merge remote-tracking branch 'upstream/master' into diffbin

mehtamansi29 committed:
Merge branch 'diffbin' of https://github.com/mehtamansi29/keras-hub into diffbin

mehtamansi29 committed:
Merge branch 'keras-team:master' into diffbin

mehtamansi29 committed:
diffbin_backbone changes

mehtamansi29 committed:
Merge remote-tracking branch 'upstream/master' into diffbin

mehtamansi29 committed:
Merge branch 'diffbin' of https://github.com/mehtamansi29/keras-hub into diffbin

mehtamansi29 committed:
diffbin_backbone fpn model changes

mehtamansi29 committed:
diffbin_textdetector description changes

mehtamansi29 committed:
Merge branch 'master' into diffbin

mehtamansi29 committed:
Fixed bracket issue

mehtamansi29 committed:
running pre-commiy

mehtamansi29 commented:
Hi @sachinprasadhs - Can you please review this PR and let me know if any changes required ?

mehtamansi29 commented:
Hi @divyashreepathihalli- Can you please review this model and let me know if any changes required ?

laxmareddyp reviewed (changes requested):
Hey @mehtamansi29 Sorry for delaying in response. 

I have a quick question on this model implementation licensing part..!

The official reference implementation (MhLiao/DB) has no license file. Under default copyright law, this means "all rights reserved" — the pretrained weights cannot be legally redistributed.

Before we proceed with the full code review, we need clarity on the licensing plan:

- Where are the pretrained weights coming from? The official MhLiao/DB repo (Google Drive)? MMOCR? Somewhere else?
- If from MhLiao/DB — do you have permission from the original authors to redistribute the converted weights on Kaggle under kaggle://keras/diffbin/...?
- If not — would you be open to using MMOCR's DBNet checkpoints instead? MMOCR is Apache 2.0 licensed, so weight redistribution is safe.
- Alternatively one option is that will you be able  to train new weights from scratch?

This is a blocker because redistributing unlicensed weights under the keras Kaggle org is not accepted. The architecture code in this PR is fine , but the weights need a clear open-source license trail.

Once we have clear understanding on the above part, we will take a decision and move further.

Thanks.
